### PR TITLE
improve state processor and provider docs

### PR DESCRIPTION
**Pull request by dunglas**
Leverage generics and autowiring attributes.

### CODE REVIEW

File: core/state-processors.md [hunk 1]
```diff
@@ -156,3 +136,27 @@ use App\State\UserProcessor;
...
+
+    App\State\UserProcessor:
+        arguments:
+            $persistProcessor: '@api_platform.doctrine.orm.state.persist_processor'
+            $removeProcessor: '@api_platform.doctrine.orm.state.remove_processor'
+            # If you're using Doctrine MongoDB ODM, you can use the following code:
+            # $persistProcessor: '@api_platform.doctrine_mongodb.odm.state.persist_processor'
+            # $removeProcessor: '@api_platform.doctrine_mongodb.odm.state.remove_processor'
+            $mailer: '@mailer'
+        tags: [ 'api_platform.state_processor' ]
```
Comment (soyuka) [inline core/state-processors.md:161]: you should add the `key` as it is used for our service locator:

```
        <service id="api_platform.state_provider.locator" class="ApiPlatform\State\CallableProvider">
            <argument type="tagged_locator" tag="api_platform.state_provider" index-by="key" />
        </service>
```

File: core/state-providers.md [hunk 2]
```diff
@@ -186,3 +170,27 @@ use App\State\BookRepresentationProvider;
...
+    App\State\BlogPostProvider: ~
+        tags: [ 'api_platform.state_provider' ]
+
+# api/config/services.yaml
+services:
+    # ...
+    App\State\BookRepresentationProvider:
+        arguments:
+            $itemProvider: '@api_platform.doctrine.orm.state.item_provider'
+        tags: [ 'api_platform.state_provider' ]
```
Comment (soyuka) [inline core/state-providers.md:195]: same we should add the `key`

File: core/state-processors.md (hunk 1 above)
Comment (dunglas) [reply core/state-processors.md:165]: Are you sure that it's necessary?

File: core/state-providers.md [hunk 3]
```diff
@@ -96,9 +85,12 @@ use ApiPlatform\Metadata\Operation;
 use ApiPlatform\State\ProviderInterface;
 use ApiPlatform\Metadata\CollectionOperationInterface;
 
+/**
+ * @implements ProviderInterface<BlogPost[]>
```
Comment (bendavies) [inline core/state-providers.md:89]: BlogPost[]|BlogPost

File: core/state-providers.md [hunk 4]
```diff
@@ -96,9 +85,12 @@ use ApiPlatform\Metadata\Operation;
...
+ */
 final class BlogPostProvider implements ProviderInterface
 {
-    public function provide(@Operation $operation, array $uriVariables = [], array $context = [])
+    public function provide(Operation $operation, array $uriVariables = [], array $context = []): array
```
Comment (bendavies) [inline core/state-providers.md:93]:
```suggestion
    public function provide(Operation $operation, array $uriVariables = [], array $context = []): array|BlogPost
```

File: core/dto.md [hunk 5]
```diff
@@ -179,12 +195,15 @@ use ApiPlatform\State\ProcessorInterface;
 use App\Dto\AnotherRepresentation;
 use App\Model\Book;
 
+/**
+ * @implements ProviderInterface<Book, AnotherRepresentation>
```
Comment (soyuka) [inline core/dto.md:199]:
```suggestion
 * @implements ProcessorInterface<Book, AnotherRepresentation>
```

File: core/dto.md [hunk 6]
```diff
@@ -47,12 +50,17 @@ use ApiPlatform\Metadata\Operation;
...
+ */
 final class UserResetPasswordProcessor implements ProcessorInterface
 {
     /**
      * @param UserResetPasswordDto $data
+     *
+     * @throws NotFoundHttpException
      */
-    public function process($data, Operation $operation, array $uriVariables = [], array $context = [])
+    public function process(mixed $data, Operation $operation, array $uriVariables = [], array $context = []): User
```
Comment (soyuka) [inline core/dto.md:63]: we can't use UserResetPasswordDto instead of mixed?

Comment (dunglas) [reply core/dto.md:63]: Unfortunately, no: https://3v4l.org/691ZG

**Pull request merged**